File: Wayner App Flutter/lib/features/logs/services/audit_log_service.dart
import '../../../core/network/api_client.dart';
import '../../../core/storage/session_storage.dart';
import '../models/audit_log.dart';

class AuditLogService {
  final ApiClient _apiClient;

  AuditLogService({ApiClient? apiClient})
      : _apiClient = apiClient ?? ApiClient();

  Future<List<AuditLog>> listarLogs({
    int limit = 100,
    int offset = 0,
    String? accion,
    String? modulo,
    String? nombreUsuario,
    String? desde,
    String? hasta,
  }) async {
    final user = await SessionStorage.getUser();

    if (user == null || user.rol.trim().toUpperCase() != 'ADMIN') {
      throw Exception('Solo el administrador puede consultar los logs');
    }

    final query = <String, dynamic>{
      'usuario_solicitante_id': user.id,
      'limit': limit,
      'offset': offset,
    };

    if (accion != null && accion.trim().isNotEmpty) {
      query['accion'] = accion.trim();
    }

    if (modulo != null && modulo.trim().isNotEmpty) {
      query['modulo'] = modulo.trim();
    }

    if (nombreUsuario != null && nombreUsuario.trim().isNotEmpty) {
      query['nombre_usuario'] = nombreUsuario.trim();
    }

    if (desde != null && desde.trim().isNotEmpty) {
      query['desde'] = desde.trim();
    }

    if (hasta != null && hasta.trim().isNotEmpty) {
      query['hasta'] = hasta.trim();
    }

    final response = await _apiClient.get('/api/logs', queryParameters: query);
    final rawData = response['data'];

    List<dynamic> data;

    if (rawData is List) {
      data = rawData;
    } else if (rawData is Map && rawData['items'] is List) {
      data = rawData['items'] as List<dynamic>;
    } else {
      data = [];
    }

    return data
        .map((item) => AuditLog.fromJson(Map<String, dynamic>.from(item)))
        .toList();
  }

  Future<void> registrarEvento({
    required String accion,
    required String modulo,
    required String detalle,
  }) async {
    final user = await SessionStorage.getUser();

    final body = {
      'usuario_id': user?.id,
      'nombre_usuario': user?.nombreUsuario,
      'rol': user?.rol,
      'accion': accion,
      'modulo': modulo,
      'detalle': detalle,
    };

    try {
      await _apiClient.post('/api/logs', body: body);
    } catch (_) {
      // El log frontend no debe romper impresión, navegación ni flujos principales.
    }
  }
}

File: Wayner Backend/app/services/audit_log_service.py
from __future__ import annotations

from app.repositories.audit_log_repository import AuditLogRepository


class AuditLogService:
    def __init__(self, repository: AuditLogRepository):
        self.repository = repository

    def create_log(
        self,
        usuario_id: int | None,
        nombre_usuario: str | None,
        rol: str | None,
        accion: str,
        modulo: str,
        metodo: str | None = None,
        ruta: str | None = None,
        estado_http: int | None = None,
        detalle: str | None = None,
        ip: str | None = None,
        user_agent: str | None = None,
        duracion_ms: int | None = None,
    ) -> int:
        return self.repository.create_log(
            usuario_id=usuario_id,
            nombre_usuario=nombre_usuario,
            rol=rol,
            accion=accion,
            modulo=modulo,
            metodo=metodo,
            ruta=ruta,
            estado_http=estado_http,
            detalle=detalle,
            ip=ip,
            user_agent=user_agent,
            duracion_ms=duracion_ms,
        )

    def list_logs(
        self,
        usuario_solicitante_id: int,
        limit: int = 100,
        offset: int = 0,
        accion: str | None = None,
        modulo: str | None = None,
        nombre_usuario: str | None = None,
        desde: str | None = None,
        hasta: str | None = None,
    ):
        self._validar_admin(usuario_solicitante_id)

        return self.repository.list_logs(
            limit=limit,
            offset=offset,
            accion=accion,
            modulo=modulo,
            nombre_usuario=nombre_usuario,
            desde=desde,
            hasta=hasta,
        )

    def count_logs(
        self,
        usuario_solicitante_id: int,
        accion: str | None = None,
        modulo: str | None = None,
        nombre_usuario: str | None = None,
        desde: str | None = None,
        hasta: str | None = None,
    ) -> int:
        self._validar_admin(usuario_solicitante_id)

        return self.repository.count_logs(
            accion=accion,
            modulo=modulo,
            nombre_usuario=nombre_usuario,
            desde=desde,
            hasta=hasta,
        )

    def cleanup_old_logs(
        self,
        usuario_solicitante_id: int,
        fecha_limite: str,
        dry_run: bool = True,
    ) -> int:
        self._validar_admin(usuario_solicitante_id)

        if dry_run:
            return self.repository.count_older_than(fecha_limite)

        return self.repository.delete_older_than(fecha_limite)

    def _validar_admin(self, usuario_solicitante_id: int) -> None:
        rol = self.repository.get_user_role(usuario_solicitante_id)

        if not rol or rol.upper() != "ADMIN":
            raise PermissionError("Solo el administrador puede consultar los logs")